pkg/yurthub/certificate/token/token.go

```diff
@@ -0,0 +1,447 @@
+/*
+Copyright 2020 The OpenYurt Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package token
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	certificatesv1 "k8s.io/api/certificates/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/authentication/user"
+	clientset "k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	certutil "k8s.io/client-go/util/cert"
+	"k8s.io/client-go/util/certificate"
+	"k8s.io/klog/v2"
+
+	"github.com/openyurtio/openyurt/cmd/yurthub/app/config"
+	"github.com/openyurtio/openyurt/pkg/projectinfo"
+	yurtutil "github.com/openyurtio/openyurt/pkg/util"
+	certfactory "github.com/openyurtio/openyurt/pkg/util/certmanager/factory"
+	"github.com/openyurtio/openyurt/pkg/util/certmanager/store"
+	kubeconfigutil "github.com/openyurtio/openyurt/pkg/util/kubeconfig"
+	"github.com/openyurtio/openyurt/pkg/util/token"
+	hubCert "github.com/openyurtio/openyurt/pkg/yurthub/certificate"
+	"github.com/openyurtio/openyurt/pkg/yurthub/util"
+)
+
+const (
+	YurtHubCSROrg           = "openyurt:yurthub"
+	defaultRootDir          = "/var/lib"
+	hubPkiDirName           = "pki"
+	hubCaFileName           = "ca.crt"
+	bootstrapConfigFileName = "bootstrap-hub.conf"
+)
+
+var (
+	hubConfigFileName                = fmt.Sprintf("%s.conf", projectinfo.GetHubName())
+	serverCertNotReadyError          = errors.New("hub server certificate")
+	apiServerClientCertNotReadyError = errors.New("APIServer client certificate")
+	caCertIsNotReadyError            = errors.New("ca.crt file")
+)
+
+type yurtHubCertManager struct {
+	client                     clientset.Interface
+	remoteServers              []*url.URL
+	caCertHashes               []string
+	apiServerClientCertManager certificate.Manager
+	hubServerCertManager       certificate.Manager
+	apiServerClientCertStore   certificate.FileStore
+	serverCertStore            certificate.FileStore
+	hubRunDir                  string
+	hubName                    string
+	joinToken                  string
+	dialer                     *util.Dialer
+}
+
+// NewYurtHubCertManager new a YurtCertificateManager instance
+func NewYurtHubCertManager(client clientset.Interface, cfg *config.YurtHubConfiguration, stopCh <-chan struct{}) (hubCert.YurtCertificateManager, error) {
```

Comment:
Is the first parameter `client` just used for unit test?

Reply:
yes, the first parameter is only used in the unit tests at present.

Reply:
So could we remove such parameter, in unit test we can change it to
```go
newMgr, err := NewYurtHubCertManager(&config.YurtHubConfiguration{
		NodeName:                 nodeName,
		RemoteServers:            remoteServers,
		CertIPs:                  certIPs,
		RootDir:                  rootDir,
		JoinToken:                joinToken,
		YurtHubCertOrganizations: []string{"yurthub:tenant:foo"},
	}, stopCh)
yurtMgr, _ := newMgr.(*yurtHubCertManager)
yurtMgr.client = client
```

Reply:
@rambohe-ch what do you think?

Reply:
@Congrool Not only unit tests under pkg/yurthub/certificate/token have used this parameter, but also other unit tests like pkg/yurthub/kubernetes/rest have used this parameter, so these unit tests can not configure `yurtMgr.client`.

BTW i will improve NewYurtHubCertManager in another pull request, and the parameters will be improved.

Reply:
got it